### DISCUSSION TITLE
numpy allclose

### DISCUSSION THREAD

**Discussion by FrancescElies** [microsoft/pyright#4536] [Q&A] [answered]
Given this script
```python
# pyright: strict
import numpy as np

np.allclose(np.array([1.0, 2.0, 3.0]), np.array([1.0, 2.0, 3.0]))

```

Mypy seems ok
```
❯ mypy --strict allclose.py
Success: no issues found in 1 source file
```

But pyright doesn't come to terms with the script above
```
❯ pyright allclose.py
No configuration file found.
No pyproject.toml file found.
stubPath C:\s\mytest\numpy\typings is not a valid directory.
Assuming Python platform Windows
Searching for source files
Found 1 source file
pyright 1.1.291
C:\s\mytest\numpy\allclose.py
  C:\s\mytest\numpy\allclose.py:4:1 - error: Type of "allclose" is partially unknown
    Type of "allclose" is "(a: _SupportsArray[dtype[Unknown]] | _NestedSequence[_SupportsArray[dtype[Unknown]]] | bool | int | float | complex | str | bytes | _NestedSequence[bool | int | float | complex | str | bytes], b: _SupportsArray[dtype[Unknown]] | _NestedSequence[_SupportsArray[dtype[Unknown]]] | bool | int | float | complex | str | bytes | _NestedSequence[bool | int | float | complex | str | bytes], rtol: float = ..., atol: float = ..., equal_nan: bool = ...) -> bool" (reportUnknownMemberType)
1 error, 0 warnings, 0 informations
Completed in 0.931sec
```

Stubfiles provides the following info.
```python
def allclose(
    a: ArrayLike,
    b: ArrayLike,
    rtol: float = ...,
    atol: float = ...,
    equal_nan: bool = ...,
) -> bool: ...

...

_ArrayLike = Union[
    _SupportsArray["dtype[_ScalarType]"],
    _NestedSequence[_SupportsArray["dtype[_ScalarType]"]],
]

```

I guess it has to do with `dtype[Unknown]` would here the right strategy be ignore `reportUnknownMemberType` in all files where `allclose` is used?

I am not sure I understand what is unknown to pyright, is this a bug or I misunderstand something, probably the latter.

**Comment by erictraut** [comment, marked as answer]
You've enabled pyright's strict mode. In this mode, the `reportUnknownMemberType` check is enabled. This check reports the use of any member (attribute or method) whose type is `Unknown` or contains an `Unknown` somewhere within it.

In this case, the `allclose` method uses the type alias `ArrayLike` which is defined as follows:
```python
ArrayLike = _DualArrayLike[
    dtype,
    Union[bool, int, float, complex, str, bytes],
]
```

The type `dtype` is a generic class that accepts a single type argument. Since none is provided, its type is `dtype[Unknown]`. That's the origin of the `Unknown` type that pyright is telling you about here. This isn't a bug in pyright. It's an omission in the numpy code, and it would need to be addressed there.
